skyportal/facility_apis/swift.py

```diff
@@ -0,0 +1,205 @@
+API_URL = f"{cfg['app.swift_protocol']}://{cfg['app.swift_host']}:{cfg['app.swift_port']}/toop/submit_json.php"
+
+
+class UVOTRequest:
+
+    """A JSON structure for Swift UVOT requests."""
+
+    def __init__(self, request):
+        """Initialize UVOT request.
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+
+        """
+
+        self.requestgroup = self._build_payload(request)
+
+    def _build_payload(self, request):
+        """Payload json for Swift UVOT TOO requests.
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+
+        Returns
+        -------
+        payload: json
+            payload for requests.
+        """
+
+        altdata = request.allocation.altdata
+
+        too = Swift_TOO()
+        too.username = altdata["username"]
+        too.shared_secret = altdata["secret"]
+        # too.debug = True
+
+        too.source_name = request.obj.id
+        too.ra, too.dec = request.obj.ra, request.obj.dec
+
+        too.source_type = request.payload["source_type"]
+
+        too.exp_time_per_visit = request.payload["exposure_time"]
+        too.monitoring_freq = "%d days" % request.payload["monitoring_freq"]
+        too.num_of_visits = int(request.payload["exposure_counts"])
+
+        too.xrt_countrate = request.payload["xrt_countrate"]
+        too.exp_time_just = request.payload["exp_time_just"]
+        too.immediate_objective = request.payload["immediate_objective"]
+
+        if request.payload["urgency"] not in ["1", "2", "3", "4"]:
+            raise ValueError('urgency outside of allowed bounds.')
+        too.urgency = int(request.payload["urgency"])
+
+        if request.payload["obs_type"] not in [
+            "Spectroscopy",
+            "Light Curve",
+            "Position",
+            "Timing",
+        ]:
+            raise ValueError('obs_type not an allowed value.')
+        too.obs_type = request.payload["obs_type"]
+
+        too.uvot_mode = request.payload["uvot_mode"]
+        too.science_just = request.payload["science_just"]
+
+        return too
+
+
+class UVOTAPI(FollowUpAPI):
+
+    """An interface to Swift operations."""
+
+    # subclasses *must* implement the method below
+    @staticmethod
+    def submit(request):
+
+        """Submit a follow-up request to Swift's UVOT
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+        """
+
+        from ..models import FacilityTransaction, DBSession
+
+        altdata = request.allocation.altdata
+        if not altdata:
+            raise ValueError('Missing allocation information.')
+
+        swiftreq = UVOTRequest(request)
+
+        swiftreq.requestgroup.validate()
+
+        r = requests.post(
```

Comment:
Handle exceptions gracefully. https://docs.python-requests.org/en/latest/user/quickstart/#errors-and-exceptions

Reply:
added